client/asset/eth/eth.go

```diff
@@ -668,12 +619,166 @@ func (eth *ExchangeWallet) unlockFunds(coins asset.Coins) error {
+	return r.expiration
+}
+
+// Coin returns the coin used to fund the swap.
+func (r *swapReceipt) Coin() asset.Coin {
+	return &coin{
+		value: r.value,
+		id:    dex.Bytes(r.txHash[:]),
+	}
+}
+
+// Contract returns the swap's secret hash.
+func (r *swapReceipt) Contract() dex.Bytes {
+	return dex.Bytes(r.secretHash[:])
+}
+
+// String returns a string representation of the swapReceipt.
+func (r *swapReceipt) String() string {
+	return fmt.Sprintf("tx hash: %x, secret hash: %x", r.txHash, r.secretHash)
+}
+
+// SignedRefund returns an empty byte array. ETH does not support a pre-signed
+// redeem script becuase the nonce needed in the transaction can not be previously
+// determined. We will need to come up with a different way to support manual
+// refunds in ETH.
+func (*swapReceipt) SignedRefund() dex.Bytes {
+	return dex.Bytes{}
+}
+
+var _ asset.Receipt = (*swapReceipt)(nil)
+
+// Swap sends the swaps in a single transaction. The Receipts returned can
+// be used to refund a failed transaction. The fees used returned are the max
+// fees that will possibly be used, since in ethereum with EIP-1559 we cannot
+// know exactly how much fees will be used.
+func (eth *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint64, error) {
+	fail := func(err error) ([]asset.Receipt, asset.Coin, uint64, error) {
+		return nil, nil, 0, err
+	}
+
+	initiations := make([]dexeth.ETHSwapInitiation, 0, len(swaps.Contracts))
+	receipts := make([]asset.Receipt, 0, len(swaps.Contracts))
+
+	eth.lockedFundsMtx.Lock()
+	defer eth.lockedFundsMtx.Unlock()
+
+	var totalInputValue uint64
+	for _, input := range swaps.Inputs {
+		if _, ok := eth.lockedFunds[input.ID().String()]; !ok {
+			return fail(fmt.Errorf("Swap: attempting to use coin that was not locked"))
+		}
+		totalInputValue += input.Value()
+	}
+
+	var totalContractValue uint64
+	secretHashMap := make(map[string]bool)
+	for _, contract := range swaps.Contracts {
+		totalContractValue += contract.Value
+
+		if len(contract.SecretHash) != srveth.SecretHashSize {
+			return fail(
+				fmt.Errorf("Swap: expected secret hash of length %v, but got %v",
+					srveth.SecretHashSize, len(contract.SecretHash)))
+		}
+		encSecretHash := hex.EncodeToString(contract.SecretHash)
+		if _, ok := secretHashMap[encSecretHash]; ok {
+			return fail(fmt.Errorf("Swap: secret hashes must be unique"))
+		}
+		secretHashMap[encSecretHash] = true
+		var secretHash [32]byte
+		copy(secretHash[:], contract.SecretHash)
+
+		if !common.IsHexAddress(contract.Address) {
+			return fail(fmt.Errorf("Swap: address in contract is not valid %v", contract.Address))
+		}
+		address := common.HexToAddress(contract.Address)
+
+		initiations = append(initiations,
+			dexeth.ETHSwapInitiation{
+				RefundTimestamp: big.NewInt(0).SetUint64(contract.LockTime),
+				SecretHash:      secretHash,
+				Participant:     address,
+				Value:           big.NewInt(0).SetUint64(contract.Value),
```

Comment:
I think this need to be converted to wei from gwei here correct? I believe all uint64 values the server deals with are gwei, and the actual contract wants the basic unit.

Reply:
Yes this was wrong.